packages/shared/lib/services/connection.service.ts

```diff
@@ -1742,6 +1742,90 @@ class ConnectionService {
+                        [targetDate, targetDate]
+                    )
+                );
+            })
+            .with('billable_connections', (qb) => {
+                qb.select(
+                    'e.account_id',
+                    'c.id as connection_id',
+                    'c.created_at',
+                    db.readOnly.raw(`COALESCE(c.deleted_at, (SELECT month_end FROM month_info) + INTERVAL '1 day') AS effective_end_date`),
+                    db.readOnly.raw(`(SELECT month_start FROM month_info) AS month_start`),
+                    db.readOnly.raw(`(SELECT month_end FROM month_info) AS month_end`),
+                    db.readOnly.raw(`(SELECT total_seconds_in_month FROM month_info) AS total_seconds_in_month`)
+                )
+                    .from('nango._nango_connections as c')
+                    .join('nango._nango_environments as e', 'c.environment_id', 'e.id')
+                    .where((builder) => {
+                        builder.where('c.deleted_at', null).orWhereRaw(`c.deleted_at >= (SELECT month_start FROM month_info)`);
+                    })
+                    .whereRaw(`c.created_at <= (SELECT month_end FROM month_info) + INTERVAL '1 day'`);
+            })
+            .with('prorated', (qb) => {
+                qb.select(
+                    'account_id',
+                    'connection_id',
+                    db.readOnly.raw(`
+                        CASE
+                            WHEN created_at < month_start AND (effective_end_date > month_end OR effective_end_date IS NULL)
+                                THEN 1.0
+                            WHEN created_at < month_start AND effective_end_date <= month_end
+                                THEN EXTRACT(EPOCH FROM (effective_end_date - month_start)) / total_seconds_in_month
+                            WHEN created_at >= month_start AND (effective_end_date > month_end OR effective_end_date IS NULL)
+                                THEN EXTRACT(EPOCH FROM (month_end + INTERVAL '1 day' - created_at)) / total_seconds_in_month
+                            ELSE EXTRACT(EPOCH FROM (effective_end_date - created_at)) / total_seconds_in_month
+                        END AS connection_weight
+                    `)
+                ).from('billable_connections');
+            })
+            .with('totals', (qb) => {
+                qb.select(
+                    'account_id as accountId',
+                    db.readOnly.raw(`FLOOR(SUM(connection_weight)) as count`),
+                    db.readOnly.raw(`${year} as year`),
+                    db.readOnly.raw(`${month} as month`)
+                )
+                    .from('prorated')
+                    .groupBy('account_id');
+            })
+            .select('*')
+            .from('totals')
+            .where('count', '>', 0);
```

Comment:
FYI: query is currently taking ~500ms against main db replica

Reply:
just me: it's definitely a case where the query builder is definitely less readable and impossible to debug without logging it first